test/unit/org/apache/cassandra/index/sai/cql/StorageAttachedIndexDDLTest.java

```diff
@@ -278,43 +326,757 @@ public void shouldFailCreationOnMultipleColumns()
+        .hasMessageContaining("Index 'index_1' already exists");
 
         // different name, same option
         assertThatThrownBy(() -> executeNet("CREATE CUSTOM INDEX index_2 ON %s(v1) USING 'StorageAttachedIndex'"))
-                .isInstanceOf(InvalidQueryException.class)
-                .hasMessageContaining("Index index_2 is a duplicate of existing index index_1");
+        .isInstanceOf(InvalidQueryException.class)
+        .hasMessageContaining("Index index_2 is a duplicate of existing index index_1");
 
-        execute("INSERT INTO %s (id, v1) VALUES(1, '1')");
+        // different name, different option, same target.
+        assertThatThrownBy(() -> executeNet("CREATE CUSTOM INDEX ON %s(v1) USING 'StorageAttachedIndex' WITH OPTIONS = { 'case_sensitive' : true }"))
+        .isInstanceOf(InvalidQueryException.class)
+        .hasMessageContaining("Cannot create more than one storage-attached index on the same column: v1" );
 
         ResultSet rows = executeNet("SELECT id FROM %s WHERE v1 = '1'");
         assertEquals(1, rows.all().size());
     }
 
+    @Test
+    public void shouldIndexBuildingWithInMemoryData() throws Throwable
+    {
+        createTable(CREATE_TABLE_TEMPLATE);
+
+        int rowCount = 10;
+        for (int i = 0; i < rowCount; i++)
+            execute("INSERT INTO %s (id1, v1) VALUES ('" + i + "', '0')");
+
+        createIndex(String.format(CREATE_INDEX_TEMPLATE, "v1"));
+        waitForIndexQueryable();
+
+        ResultSet rows = executeNet("SELECT id1 FROM %s WHERE v1='0'");
+        assertEquals(rowCount, rows.all().size());
+    }
+
+    @Test
+    public void shouldIndexExistingMemtableOnCreationWithConcurrentFlush() throws Throwable
+    {
+        createTable("CREATE TABLE %s (id text PRIMARY KEY, val text)");
+        execute("INSERT INTO %s (id, val) VALUES ('1', 'Camel')");
+
+        Injections.Barrier delayInitializationTask =
+        Injections.newBarrier("delayInitializationTask", 2, false)
+                  .add(InvokePointBuilder.newInvokePoint().onClass(StorageAttachedIndex.class).onMethod("startInitialBuild"))
+                  .build();
+
+        // Create the index, but do not allow the initial index build to begin:
+        Injections.inject(delayInitializationTask);
+        createIndex("CREATE CUSTOM INDEX ON %s(val) USING 'StorageAttachedIndex'");
+
+        // Flush the Memtable's contents, which will feed data to the index as the SSTable is written:
+        flush();
+
+        // Allow the initialization task, which builds the index, to continue:
+        delayInitializationTask.countDown();
+
+        waitForIndexQueryable();
+
+        ResultSet rows = executeNet("SELECT id FROM %s WHERE val = 'Camel'");
+        assertEquals(1, rows.all().size());
+
+        assertEquals("Segment memory limiter should revert to zero.", 0L, getSegmentBufferUsedBytes());
+        assertEquals("There should be no segment builders in progress.", 0L, getColumnIndexBuildsInProgress());
```

Comment:
nit: We check these two things in many tests. Might be worth a helper method, or not...